strawberryfields/api/devicespec.py

```diff
@@ -15,62 +15,66 @@
 This module contains a class that represents the specifications of
 a device available via the API.
 """
-from collections.abc import Sequence
-import re
+from typing import Sequence, Mapping, Any
 
 import blackbird
 from blackbird.error import BlackbirdSyntaxError
 
 import strawberryfields as sf
 from strawberryfields.compilers import Ranges
-from strawberryfields.tdm.tdmprogram import TDMProgram
 
 
 class DeviceSpec:
     """The specifications for a specific hardware device.
 
     Args:
-        target (str): name of the target hardware device
         spec (dict): dictionary representing the raw device specification.
-            This dictionary should contain the following key-value pairs:
+            This dictionary must contain the following key-value pairs:
 
-            - layout (str): string containing the Blackbird circuit layout
+            - layout (str): string containing the Blackbird or XIR circuit layout
             - modes (int): number of modes supported by the target
             - compiler (list): list of supported compilers
             - gate_parameters (dict): parameters for the circuit gates
-
-        connection (strawberryfields.api.Connection): connection over which the
-            job is managed
     """
 
-    def __init__(self, target, spec, connection):
-        self._target = target
-        self._connection = connection
+    def __init__(self, spec: Mapping[str, Any]) -> None:
+        if not {"layout", "modes", "compiler", "gate_parameters"} <= spec.keys():
+            missing_keys = {
+                "target",
+                "layout",
+                "modes",
+                "compiler",
+                "gate_parameters",
+            } - spec.keys()
+            raise ValueError(
+                f"Device specification is missing the following keys: {sorted(missing_keys)}"
+            )
+
         self._spec = spec
 
     @property
-    def target(self):
+    def target(self) -> str:
         """str: The name of the target hardware device."""
-        return self._target
+        return self._spec["target"]
 
     @property
-    def layout(self):
+    def layout(self) -> str:
         """str: Returns a string containing the Blackbird circuit layout."""
         return self._spec["layout"]
 
     @property
-    def modes(self):
+    def modes(self) -> int:
         """int: Number of modes supported by the device."""
         return self._spec["modes"]
 
     @property
-    def compiler(self):
+    def compiler(self) -> Sequence[str]:
```

Comment:
I noticed in some of the RemoteEngine tests that `DeviceSpec.compiler` can also just be a string. Is that true?

Reply:
`_spec["compiler"]` should always be a list of compiler as far as I'm aware. `_spec["default_compiler"]` is a single compiler (the first in the list). Could you point to anywhere where `_spec["compiler"]` is a string?

Reply:
Sure! The tests I was referring to are:
* [`TestRemoteEngineIntegration.test_recompilation_run_async`](https://github.com/XanaduAI/strawberryfields/blob/integrate-xanadu-cloud-client/tests/api/test_remote_engine.py#L261)
* [`TestRemoteEngineIntegration.test_recompilation_run`](https://github.com/XanaduAI/strawberryfields/blob/integrate-xanadu-cloud-client/tests/api/test_remote_engine.py#L280)
* [`TestRemoteEngineIntegration.test_validation`](https://github.com/XanaduAI/strawberryfields/blob/integrate-xanadu-cloud-client/tests/api/test_remote_engine.py#L303)

Reply:
Just updated them to be lists instead. The tests pass either way, and they should be lists as far as I'm concerned. 🙂